lib/elixir/lib/code/normalizer.ex

```diff
@@ -0,0 +1,246 @@
+    args =
+      Enum.map(args, fn
+        arg when is_list(arg) ->
+          {_, _, [arg]} = do_normalize(arg, meta)
+          arg
+
+        arg ->
+          do_normalize(arg, meta)
+      end)
+
+    {:when, meta, args}
+  end
+
+  # Calls
+  defp do_normalize({form, meta, args}, _parent_meta) when is_list(args) do
+    # Only normalize the form if it's a qualified call
+    form =
+      if is_atom(form) do
+        form
+      else
+        do_normalize(form, meta)
+      end
+
+    cond do
+      Keyword.has_key?(meta, :do) ->
+        {last_arg, leading_args} = List.pop_at(args, -1)
+
+        last_arg =
+          Enum.map(last_arg, fn {tag, block} ->
+            block = do_normalize(block, meta)
+
+            block =
+              case block do
+                {_, _, [[{:->, _, _} | _] = block]} -> block
+                block -> block
+              end
+
+            # Only wrap the tag if it isn't already wrapped
+            tag =
+              case tag do
+                {:__block__, _, _} -> tag
+                _ -> {:__block__, [line: meta[:line]], [tag]}
+              end
+
+            {tag, block}
+          end)
+
+        {_, _, [leading_args]} = do_normalize(leading_args, meta)
+
+        {form, meta, leading_args ++ [last_arg]}
+
+      true ->
+        args = Enum.map(args, &do_normalize(&1, meta))
+
+        {form, meta, args}
+    end
+  end
+
+  # Integers, floats, atoms
+  defp do_normalize(x, parent_meta) when is_literal(x) do
+    meta = [line: parent_meta[:line]]
+
+    meta =
+      if not is_atom(x) do
+        Keyword.put(meta, :token, Macro.to_string(x))
+      else
+        meta
+      end
+
+    meta =
+      if not is_nil(parent_meta[:format]) do
+        Keyword.put(meta, :format, parent_meta[:format])
+      else
+        meta
+      end
+
+    {:__block__, meta, [x]}
+  end
+
+  # 2-tuples
+  defp do_normalize({left, right}, parent_meta) do
+    meta = [line: parent_meta[:line]]
+
+    left_parent_meta =
+      if is_atom(left) do
+        Keyword.put(parent_meta, :format, :keyword)
+      else
+        meta
+      end
+
+    {:__block__, meta,
+     [
+       {do_normalize(left, left_parent_meta), do_normalize(right, parent_meta)}
+     ]}
+  end
+
+  # Lists
+  defp do_normalize(list, parent_meta) when is_list(list) do
+    if !Enum.empty?(list) and List.ascii_printable?(list) do
+      # It's a charlist
+      {:__block__, [line: parent_meta[:line], delimiter: "'"], [list]}
+    else
+      meta = [line: parent_meta[:line], closing: [line: parent_meta[:line]]]
+
+      args = normalize_list_elements(list, parent_meta)
+
+      {:__block__, meta, [args]}
+    end
+  end
+
+  # Everything else
+  defp do_normalize(quoted, _parent_meta) do
+    quoted
+  end
+
+  defp normalize_list_elements(elems, parent_meta, keyword? \\ false)
+
+  defp normalize_list_elements([[{_, _, [{_, _}]}] = first | rest], parent_meta, keyword?) do
+    # Skip already normalized 2-tuples
```

Comment:
Why would we need this? It seems we might be normalizing some things twice?

Reply:
If we have a list like this one: `[1, {:foo, :bar}, 2, a: :b, c: :d]`, the `{:foo, :bar}` needs to be wrapped in a block because it's a 2-tuple literal, but for the last two only their contents need to be wrapped (`{{:__block__, [format: :keyword], [:key]}, {:__block__, [], [:value]}}`), but I can't decide when to do it without checking if the current element is part of a keyword list or not.